### ISSUE TITLE
BTD Inversion

### ISSUE THREAD

**Issue by AleksBL** [dipc-cc/hubbard#101]
Hi, 
Have worked a bit on the BTD inversion. One problem there is right now is the for-loops over energy in the NEGF code, which makes the code i have a bit hard to implement in your code. I worked a bit on a workaround, but it didnt work out right now. 

In the attached file is a really barebones implementation for a BTD inversion (iG_dropin.py, the use of which is demonstrated in the bottom of the Test_Ribbon.py script)... for the example on a 1440 orbital Hamiltonian it is twice as fast as the scipy implementation. There is a lot of room to make it faster, and the secondlast block in the Test_Ribbon.py script hopefully demonstrates this, but we need to make the inversions in one go instead as this would remove a lot of overhead converting between sparse , block matrix and numpy formats.... :) The code can approach (but still a bit slower than) tbtrans speeds if larger batches of energy-points are used. 

The Test_Ribbon script  demonstrates the uses, and has an outcommented part where the DFT calculation is made.
[Send.tar.gz](https://github.com/dipc-cc/hubbard/files/6836496/Send.tar.gz)

BR. Aleksander

**Comment by zerothi** [comment]
FYI did you see the btd in sisl/toolbox/btd? If your code enhances that one it could be of general use? ;-)

**Comment by sofiasanz** [comment]
Hi Aleksander, thank you for the work! :)

I see. I will take a look at the code. Probably is not so hard to avoid the energy loops by inverting the multidimensional array at once. I think that if we use `numpy.linalg.inv` it can invert this type of matrix.

BTW I have seen that you import some python modules, such as `seek path` that obtains band paths in the Brillouin zone of crystal structures. Maybe this could be done with `sisl`? Just to avoid having to import too many modules.

**Comment by AleksBL** [comment]
Yes, the imports can be cut down to numpy, scipy, sisl and optionally numba, a lot of the other imported stuff is superfluous for this. :)

**Comment by sofiasanz** [comment]
Hi! I made a few changes in #105 following your comment (https://github.com/dipc-cc/hubbard/issues/101#issue-947058764). There I eliminated the energy for loops to obtain the Green's function per energy point, instead the Green's function now is obtained in one go for all energy points. Is something like this what you expected to have in order to implement your `BTD` code?

**Comment by AleksBL** [comment]
Yes, this is more compatible!
I have a working _G right now, but i also want it to use the fact that the greens function is a symmetric matrix(and only calculate half the matrix elements), so ill just send you the code soon'ish... :)

**Comment by AleksBL** [comment]
Take a look at the "Calculation.py in this one.... Cut out a lot of crap from the code hopefully the "block_linalg" code is better. :) There might be some functions that arent used for this that dont work.
[Send2.zip](https://github.com/dipc-cc/hubbard/files/7111390/Send2.zip)

**Comment by AleksBL** [comment]
@sofiasanz by the way, is the elec_idx arrays supposed to have the (1,n) shape, or is it me who have gotten the wrong impression at some point? :)

**Comment by zerothi** [comment]
> @sofiasanz by the way, is the elec_idx arrays supposed to have the (1,n) shape, or is it me who have gotten the wrong impression at some point? :)

The usefulness of indices with that dimensionality can reduce redundant indices, in your code you have lots of:
```python
pvt = pvt.ravel() # just to make it clear it is 1D
G = G[pvt, :][:, pvt]
```
however, this can be done in a single step with:
```python
pvt = pvt.reshape(-1, 1)
G = G[pvt, pvt.T]
```
since it becomes b-casted to correct indices.

**Comment by AleksBL** [comment]
@zerothi I just think it introduces a hidden transposition of the self-energies if it is indeed (1,n)? if it was (n,1) i think it would be fine :) its very inconsequencial though because the selfenergies are symmetric...

**Comment by zerothi** [comment]
> @zerothi I just think it introduces a hidden transposition of the self-energies if it is indeed (1,n)? if it was (n,1) i think it would be fine :) its very inconsequencial though because the selfenergies are symmetric...

true, but generally it should be `(n,1)` I don't know about the `(1,n)` shape. But self-energies are not symmetric @ `k`.

**Comment by AleksBL** [comment]
I think its me who have made this mistake in the elec_idx to start with... :)

**Comment by AleksBL** [comment]
Hi,
Just to follow up on the BTD code; 
Whats the biggest system that have been treated using this method your code implements? And should we see if we can beat it?
I also think there's a fairly easy fix for the pivotting problem we encountered. We can just make an additional piece of code that expands the blocks where the tbt-trans self-energies lives to also just include the electrode indices....

\\ Aleksander

**Comment by sofiasanz** [comment]
Hi @AleksBL, sorry for the delay, I had been a little busy :)

I would say that the largest system that we have treated with is the device formed of crossed GNRs (around 900 atoms). And that I can run it locally in my computer with the brute force inversion methodology.

For the electrode indices, nice! Do you need anything from my side?
We can also use the pivoting indices from TBT and fold them into the device as TBT does, but only in the case of BTD inversion.

Sofia

**Comment by AleksBL** [comment]
This seems to work:

```python 
def get_btd_partition_of_matrix(_csr, start):
    from scipy.sparse.csgraph import reverse_cuthill_mckee as rcm
    from scipy import sparse as sp
    assert _csr.shape[0] == _csr.shape[1]
    p = rcm(_csr)
    no = _csr.shape[0]
    csr = _csr[p,:][:,p]
    Part = [0, start]
    x0   = Part[-1]
    while x0<no:
        #print(x0)
        sl = slice(Part[-2], Part[-1])
        coup = csr[sl,sl.stop:]
        i,j,v = sp.find(coup)
        if len(i)==0:
            break
        else:
            x0 = np.max(j)+1
            Part += [Part[-1] + x0]
    btd = [Part[i+1] - Part[i] for i in range(len(Part)-1)]
    return p, btd, Part
```

**Comment by AleksBL** [comment]
Hi again, 
scipy turned out to have this function called "reverse_cuthill_mckee" which gives a pivotting scheme for the device without throwing away the electrode indices.
Just to show some of the results from the Test_BTD3.py script:
geometry:
![image](https://user-images.githubusercontent.com/75378674/198037632-6bf24cdd-a360-4a7d-9f9d-bbeb5c0dc083.png)
with BTD:
![image](https://user-images.githubusercontent.com/75378674/198037732-04f9c072-a4e4-4405-9681-9ad8901798a7.png)
without:
![image](https://user-images.githubusercontent.com/75378674/198037863-f11ca8ea-7d7a-4b6c-a80d-dac201193588.png)

It seems to give almost the same as the regular way, but there is this small descrepancy on the decimals. I dont know if its a float32 vs float64 artifact or if they actually calculate something different. 

Larger geometry, 4500 sites, very narrow and long which is the kind of structures the BTD will perform well on. RAM usage is ~5GB max. 
![image](https://user-images.githubusercontent.com/75378674/198039515-d760e57f-e23a-4112-a46d-a43858c1ef7a.png)

![image](https://user-images.githubusercontent.com/75378674/198047564-c51691f6-1456-4ba9-a78e-d377840ec6ac.png)
